### ISSUE TITLE
Estimate monetary cost

### ISSUE THREAD

**Issue by TomNicholas** [tomwhite/cubed#334]
Cubed arguably has enough information to give a rough estimate of the monetary cost of executing the plan before starting execution.

I'm imagining a new method `.estimate_cost(executor)` that is similar to `.compute(executor)`. Calling this we would know 
* how many arrays are to be processed, how big they all are, and what numpy functions are to be used to process them,
* which serverless executor the functions are to be run on,
* the bucket information,

It would just print an estimation of the cost back to the user without running anything, and maybe raise warnings if they are planning to do something that seems obviously expensive (e.g. like having their temporary bucket for intermediate data be in AWS but their executor be GCF).

This means if we had a little table somewhere of e.g. AWS lambda and S3 prices, Cubed could consult those numbers and sum them. It would require an idea of e.g. how long it takes to run `np.mean()` on a chunk of a certain size on a certain container, but this seems like something that can be discovered fairly straightforwardly.

Obviously there are a long tail of cases where this wouldn't work, but often you might still be able to provide a lower bound cost estimate. For example if your plan had a step that applied some arbitrary function with `apply_gufunc`, you would not know if that was some super expensive function that would run for ever, but you would still be able to estimate the minimum cost assuming that that function was very light.

**Comment by tomwhite** [comment]
Duplicate of #219?

**Comment by TomNicholas** [comment]
I had forgotten about #219, but actually I don't think this is a duplicate - I'm suggesting warning users of *estimated costs before execution*, whereas #219 seems to be about *actual cost after execution*. Though I imagine you could re-use much code when calculating both numbers.

Basically I think it would be useful for users to be able to see "hang on, this isn't supposed to cost that much, maybe I've not expressed the analysis I meant to..." before they actually waste that money.

**Comment by tomwhite** [comment]
Sounds good - let's keep both open.

**Comment by TomNicholas** [comment]
@sharkinsspatial suggested an interesting use case for this feature: estimating the cost of a large computation months in advance so as to use it when budgeting for a grant proposal.

I realised you could even do that for a dataset that doesn't actually exist yet - you just have to create a virtual dataset pointing to imaginary data the same way I did in the VirtualiZarr/Icechunk performance benchmark, and ensure that the virtual variables have the same shape, dtype, and chunking as the dataset you eventually plan to compute on. The cubed Plan then could never be executed, but it's cost could be estimated.

Planning costs far in advance like this might be a motivation to ensure the cost model includes versions of packages, as e.g. and optimization in cubed might significantly change the cost months later. Not that this is intended for anything more than order-of magnitude estimates though.
